draft: doc for neuron custody projects (#2628)
dfinity/portal: docs/developer-docs/daos/nns/neuron-management.md

> ---

Reviewer: This file will need to be a `.mdx` instead of a `.md`.

Author: Thanks for the feedback. This is what I expected, but I didn't manage to make this readable in my local dev environment.. Maybe I was missing a trick.
Anyways, I thought I write it like this and convert it in the end. Let me keep this comment open as a reminder

keywords: [intermediate, governance, concept, nns]
---

import { MarkdownChipRow } from "/src/components/Chip/MarkdownChipRow";

# Neuron management

<MarkdownChipRow labels={["Intermediate", "Governance", "Concept"]} />

## Overview 

> This page describes the technical APIs and details to manage neurons on the NNS governance canister.

Reviewer: ```suggestion
This page describes the technical APIs used to manage neurons on the NNS.
```

Author: Applied in local copy and will be commited in next version.

> If you are building a project that offers neuron staking or custody support, this is the page for you.

Reviewer: ```suggestion
If you are building a project that offers neuron staking or custody support, this page is particularly useful. 
```

Author: applied in next commit

> You can find the most up-to-date `candid` definition for the NNS governance canister [here](https://github.com/dfinity/ic/blob/master/rs/nns/governance/canister/governance.did).

Reviewer: ```suggestion
View the most up-to-date [`candid` definition for the NNS governance canister](https://github.com/dfinity/ic/blob/master/rs/nns/governance/canister/governance.did).
```

Author: applied in next commit

> We will here focus on the neuron interactions and copy some of the relevant definitions from this file to explain them.

Reviewer: ```suggestion
This page will focus on the neuron interactions and explain some of the relevant definitions from this Candid file.
```

## Neuron attributes

> Before explaining how to modify a neuron, let us first understand the most important parts of a neuron.

Reviewer: ```suggestion
Before explaining how to modify a neuron, it is important to understand the most important parts of a neuron.
```

Every user can stake ICP utility tokens into a neuron and pariticipate in the NNS DAO with this neuron.  

* **Neuron ID**: Each neuron has an identity selected by NNS governance at neuron creation.
* **Account**: Each neuron has an associated account on the ICP ledger where the locked ICP balance resides. This account is owned by the NNS governance canister and therefore a user cannot move staked tokens.

> * **Controller**: The principal that controls the neuron. The principal must identify a public key pair, which acts as a “master key,” such that the corresponding secret key should be kept very secure. The principal might control many neurons.

Reviewer: ```suggestion
* **Controller**: The principal that controls the neuron. The principal must identify a public key pair, which acts as a “master key,” such that the corresponding secret key should be kept very secure. A principal can control many neurons.
```

* **Dissolve delay & dissolve state**: The tokens in a neuron are locked for a specified time, called the _dissolve delay_. This can be thought of like a timer.
  * A neuron can be _non-dissolving_ which means that the timer is stopped and the neuron's dissolve delay remains the same. 
  * A neuron can be _dissolving_ which means that the timer is decreasing the neuron's dissolve delay as time passes. 

>   * Once the timer is all the way down a neuron is _dissolved_ and the ICP tokens can be unstaked again.

Reviewer: ```suggestion
  * Once the timer has counted down, a neuron is _dissolved_ and the ICP tokens can be unstaked again.
```

* **Age**: Every non-dissolving neuron has an age, which denotes how long it has been in the non-dissolving state. 
A neuron's dissolve delay and age increases the neuron's voting power. 

> * **Maturity (positive number – percent)**:  When neurons vote, over time the NNS increases their so called maturity to reward them. Maturity can be converted into ICP by spawning (see below).

Reviewer: ```suggestion
* **Maturity (positive number – percent)**:  When neurons vote, over time the NNS increases their maturity to reward them. Maturity can be converted into ICP by spawning (see below).
```

Author: applied in next commit

### Interacting with neurons

> All interactions with a neuron go through the following API on the NNS governance cansiter. This includes everything in a neuron's lifetime such as the neuron's creation, modifications and voting, and finally unstaking the ICP tokens again from the neuron.

Reviewer: ```suggestion
All interactions with a neuron go through the following API on the NNS governance canister. This includes everything in a neuron's lifetime such as the neuron's creation, modifications and voting, and unstaking the ICP tokens.
```

Author: applied in next commit

```
type ManageNeuron = record {
  id : opt NeuronId;
  command : opt Command;
  neuron_id_or_subaccount : opt NeuronIdOrSubaccount;
};
```

> We will next to through a few important commands and how they can be used.

Reviewer: ```suggestion
```

### Creating a neuron and topping it up
Staking ICP utility tokens in a neuron involves two commands:
1. Sending ICP utility tokens to the neuron's subaccount.

> 2. Claiming the neuron on the NNS governance canister, which basically means telling the governance canister that the transfer in 1. happened upon which NNS governance will create a new neuron.

Reviewer: ```suggestion
2. Claiming the neuron on the NNS governance canister, which tells the governance canister that the transfer in (1) happened ,upon which the NNS governance will create a new neuron.
```

TODO: ADD DETAILS HOW THE SUBACCOUNT MUST BE COMPUTED!

> For the ledger transfer, the relevant `candid` interface can be found [here](https://github.com/dfinity/ic/blob/master/rs/rosetta-api/icp_ledger/ledger.did#L25) and contains the following details for a transaction:

Reviewer: ```suggestion
For the ledger transfer, view the [relevant `candid` interface](https://github.com/dfinity/ic/blob/master/rs/rosetta-api/icp_ledger/ledger.did#L25) which contains the following details for a transaction:
```

```
type Transaction = record {
    memo : Memo;
    icrc1_memo: opt blob;
    operation : opt Operation;
    created_at_time : TimeStamp;
}; 
```

> To then claim the neuron, you can use the following command:

Reviewer: ```suggestion
To claim the neuron, use the following command:
```

TODO: WHICH WAY IS RECOMMENDED? ADD ARGUMENT EXPLANATION
```
type ClaimOrRefresh = record { by : opt By };
type By = variant {
  NeuronIdOrSubaccount : record {};
  MemoAndController : ClaimOrRefreshNeuronFromAccount;
  Memo : nat64;
};
type ClaimOrRefreshNeuronFromAccount = record {
  controller : opt principal;
  memo : nat64;
};
```

> Topping up an exising neuron with more tokens is called _refreshing_ a neuron and works in the same way. In a first step, additional tokens are sent to the neuron's account. In the second step, the NNS governance cansiter is informed so that it can update the neuron's stake after confirming with the ledger that there are new tokens. Note that refreshing of a neuron will also update the neuron's age to account for the fact that the newly added tokens have no age.

Reviewer: ```suggestion
Topping up an existing neuron with more tokens is called _refreshing_ a neuron and works in the same way. 

First, additional tokens are sent to the neuron's account. Then, the NNS governance canister is informed so that it can update the neuron's stake after confirming with the ledger that there are new tokens. 

:::info
Note that refreshing of a neuron will also update the neuron's age to account for the fact that the newly added tokens have no age.
:::
```

TODO: ADD ARGUMENT EXPLANATION AND WHAT TO DO DIFFERENTLY FOR THE REFRESHING

### Modifying a neuron's state

> * **Start dissolving**: The dissolve delay is like a kitchen timer that can only be turned in one direction. It can be arbitrarily increased, but only reduced by turning on dissolve mode and counting down. The neuron can be instructed to start “dissolving.” When the neuron is dissolving, its dissolve delay falls over time until it is either stopped or reaches zero. A neuron cannot vote (or earn rewards for voting) when its dissolve delay falls below six months. Once the dissolve delay reaches zero, it stops falling and the controlling principal can instruct the neuron to disburse.

Reviewer: ```suggestion
* **Start dissolving**: The neuron dissolve delay is like a kitchen timer that can only be turned in one direction. It can be arbitrarily increased, but only reduced by turning on dissolve mode and counting down. When the neuron is dissolving, its dissolve delay falls over time until it is either stopped or reaches zero. A neuron cannot vote (or earn rewards for voting) when its dissolve delay falls below six months. Once the dissolve delay reaches zero, it stops falling and the controlling principal can instruct the neuron to disburse.
```

* **Increase dissolve delay**: The dissolve delay of a neuron can be increased up to a maximum of eight years.
```
type Configure = record { operation : opt Operation };
type Operation = variant {
  ...
  StopDissolving : record {};
  StartDissolving : record {};
  IncreaseDissolveDelay : IncreaseDissolveDelay;
  ...
  SetDissolveTimestamp : SetDissolveTimestamp;
};
type IncreaseDissolveDelay = record {
  additional_dissolve_delay_seconds : nat32;
};
type SetDissolveTimestamp = record { dissolve_timestamp_seconds : nat64 };
```

### Managing permissions of a neuron
* **Hot Keys (list of principal ID)**: Keys that can be used to perform actions with limited privileges, such as voting, without exposing the secret key corresponding to the principal (e.g., could be a WebAuthn key).

* **Add hot key**: Add a new hot key that can be used to manage the neuron. This provides an alternative to using the principal’s cold key to manage the neuron, which might be onerous and difficult to keep secure, especially if it is used regularly. A hot key might be a WebAuthn key that is maintained inside a user device, such as a smartphone.

> * **Remove hot key**: Temove a hot key that has been previously assigned to the neuron.

Reviewer: ```suggestion
* **Remove hot key**: Remove a hot key that has been previously assigned to the neuron.
```

* **Follow relationships (mapping from topic to list of followers)**: A neuron can be configured to vote automatically by following other neurons on a topic-by-topic basis. For any valid topic, a list of followers can be specified, and the neuron will follow the vote of a majority of the followers on a proposal with a type belonging to that topic. If a null topic is specified, this acts as a catch-all that enables the neuron to follow the vote of followees where a rule has not been specified.

* **Recent votes**: A record of recent votes is maintained. This can provide a guide for those wishing to evaluate whether to follow a neuron or how their followers are voting.

* **Vote**: Have the neuron vote to either adopt or reject a proposal with a specified ID.
* **Follow**: Add a rule that enables the neuron to vote automatically on proposals that belong to a specific topic, by specifying a group of followee neurons whose majority vote is followed. The configuration of such follow rules can be used to:
    - Distribute control over voting power amongst multiple entities.
    - Have a neuron vote automatically when its owner lacks time to evaluate newly submitted proposals.

>     - Have a neuron vote automatically when its own lacks the expertise to evaluate newly submitted proposals.

Reviewer: ```suggestion
    - Have a neuron vote automatically when its owner lacks the expertise to evaluate newly submitted proposals.
```

    - For other purposes.

A follow rule specifies a set of followers. Once a majority of the followers vote to adopt or reject a proposal belonging to the specified topic, the neuron votes the same way. If it becomes impossible for a majority of the followers to adopt (for example, because they are split 50–50 between adopt and reject), then the neuron votes to reject. If a rule is specified where the proposal topic is null, then it becomes a catch-all-follow rule, which will be used to vote automatically on proposals belonging to topics for which no specific rule has been specified. If the list of followers is empty, this effectively removes the following rule.

> TODO: HOW TO GET THE ID OF TOPICS BEFORE DOING THIS?

Author: @max-dfinity is this something that you can hep with? How do I learn form the outside which topic has which ID so that I can set following? I am assuming there must be an API for it?

  * A neuron can be _non-dissolving_ which means that the timer is stopped and the neuron's dissolve delay remains the same. 
  * A neuron can be _dissolving_ which means that the timer is decreasing the neuron's dissolve delay as time passes. 
  * Once the timer has counted down, a neuron is _dissolved_ and the ICP tokens can be unstaked again.
* **Age**: Every non-dissolving neuron has an age, which denotes how long it has been in the non-dissolving state. 
A neuron's dissolve delay and age increases the neuron's voting power. 

* **Maturity (positive number – percent)**:  When neurons vote, over time the NNS increases their maturity to reward them. Maturity can be converted into ICP by spawning (see below).

### Interacting with neurons
All interactions with a neuron go through the following API on the NNS governance canister. This includes everything in a neuron's lifetime such as the neuron's creation, modifications and voting, and unstaking the ICP tokens.
```
type ManageNeuron = record {
  id : opt NeuronId;
  command : opt Command;
  neuron_id_or_subaccount : opt NeuronIdOrSubaccount;
};
```

> ### Creating a neuron and topping it up

Author: REVIEW FROM NNS TEAM WANTED

Reviewer 2: Yes, consider it done!

View for example the method [`compute_neuron_staking_subaccount_bytes`](https://github.com/dfinity/ic/blob/master/rs/nervous_system/common/src/ledger.rs) which computes the subaccount given a controller principal and a nonce. Make sure you are using an up-to-date version of this code in your application.

:::info
It is of utmost importance to ensure that the subaccount is computed correctly in the first step and remembered so that it can be reused in the second step.
If this fails, funds could be sent to a dead account and be unrecoverable.

:::

First learn the controller principal and choose a nonce.
Then, for the ledger transfer, view the [relevant `candid` interface](https://github.com/dfinity/ic/blob/master/rs/rosetta-api/icp_ledger/ledger.did#L25) which contains the following details for a transaction.
```
type Transaction = record {
    memo : Memo;
    icrc1_memo: opt blob;
    operation : opt Operation;
    created_at_time : TimeStamp;
}; 
type Operation = variant {
    ...
    Transfer : record {
        from : AccountIdentifier;
        to : AccountIdentifier;
        amount : Tokens;
        fee : Tokens;
        spender : opt vec nat8;
    };
    ...
};
```

> To make a transfer to the correct subaccount, choose the `memo` equal your chosen nonce and `operation` to be a transfer. You can make the transfer from any `from` account that you control. The `to` account should be the computed subaccount for the neuron. To later stake a neuron, the `amount` must be at least 1 ICP. Set the `fee` to the standard fee of the ICP ledger canister.

Reviewer 3: So the "to" account has to be AccountIdentifier::new(GOVERNANCE_CANISTER_ID.get(), Some(subaccount))

If you look at claim_neuron you can see that neuron_subaccount is used.  This uses the computed subaccount along with Governance to create the account that is checked on the ledger. https://sourcegraph.com/github.com/dfinity/ic@0cdb6f9d14a65906f0089b6571fcdd5b9867d19b/-/blob/rs/nns/governance/src/governance.rs?L5675 

The subaccount is computed in the way you say, but I worry here that someone might try to send funds directly to the the subaccount in this step.  The subaccount is only part of the "to" field for the transfer.  The principal is GOVERNANCE_CANISTER_ID.  

What that actually looks like in bytes would be very different than the subaccount.

Author: Andre looked at this in the meantime so I am resolving the thread

### Voting with a neuron
A neuron represents a participant of the NNS DAO and can participate in governance. Concretely, a neuron can submit proposals or vote on them if its dissolve delay is at least 6 months. 
A neuron can either directly vote to adopt or reject a given proposal or it can _follow_ other neurons for certain decisions - which will automatically cast a ballot for the neuron if enough of the followees voted.

**Vote directly**
To vote directly, first find out what proposals can be voted on and then find the ID of the proposal you want to vote on.
The following are the relevant NNS governance API and record that can also be found in the candid file.
```
service : (Governance) -> {
  get_pending_proposals : () -> (vec ProposalInfo) query;
};
type ProposalInfo = record {
  id : opt NeuronId;
  ...
};
 ```
First use the `get_pending_proposals` API to get the `ProposalInfo` of all proposals that are still pending. 
Then, as part of this proposal information you can find a proposal's id `id`.

Using this, you can now vote on a proposal using the following neuron command.
```
type RegisterVote = record { 
  vote : int32; 
  proposal : opt NeuronId 
};
```

> The `vote` represents if the neuron should adopt or reject the proposal and the `proposal` is the proposal ID of the proposal that the neuron votes on.

Author: What is Yes and what is no? How can a user find out?

Reviewer 2: It is defined here: https://gitlab.com/dfinity-lab/public/ic/-/blob/7d17d2785317c6c19ecd00599872461154acf4a1/rs/nns/governance/proto/ic_nns_governance/pb/v1/governance.proto#L427-437

```
  VOTE_YES = 1;
  VOTE_NO = 2;
  ```

Reviewer 2: I don't know how a user would normally be expected to find this out

For any valid topic, a list of followees can be specified, and the neuron will follow the vote of a majority of the followers on a proposal with a type belonging to that topic. If a null topic is specified, this acts as a catch-all that enables the neuron to follow the vote of followees for all topics where no explicit rule has been specified. 
There are two topics that are excluded from the catch-all: `Governance` and `SNS & Neuron's Fund`. 

:::info
It is important to understand that a neuron's vote is only cast if the majority of the follwed neurons agree on a decision. Find [here](https://internetcomputer.org/docs/current/developer-docs/daos/nns/nns-app-quickstart#neuron-following) more information and examples regarding the following-rules.

:::

To define a follow rule, use the following neuron command
 ```
type Follow = record {
  topic : int32; 
  followees : vec NeuronId 
};
```
The argument `topic` defines for which proposal topic this following rule should be applied.
If the topic is `null`, then this rule is applied for the catch-all and will be applied for all topics where no other explicit rule is defined except for`Governance` and `SNS & Neuron's Fund`.

> The second argument `followees` defines a list of neuron IDs whose decision will be followed. There can be up to 15 followees for each topic (but more followees is not necessarily better - see the rules linked above).

Author: TODO: HOW TO GET THE ID OF TOPICS?

Reviewer 2: https://github.com/dfinity/ic/blob/7184d168d44f3ba27f750285a0fba43a0c56190a/rs/nns/governance/src/gen/ic_nns_governance.pb.v1.rs#L3142 The IDs are specified here

Author: Thanks. I now added this, would be great if you could double check whether the explanation makes sense!

Reviewer 2: For some reason I don't see where you added it 🤔 Could you point out where? (and maybe DM me so I can respond immediately)

## Overview 
This page describes the technical APIs used to manage neurons on the NNS.
If you are building a project that offers neuron staking or custody support, this page is particularly useful.

View the most up-to-date [`candid` definition for the NNS governance canister](https://github.com/dfinity/ic/blob/master/rs/nns/governance/canister/governance.did).
This page will focus on the neuron interactions and explain some of the relevant definitions from this Candid file.
To keep up to date with new NNS release proposals from DFINITY, you can subscribe to [this forum thread](https://forum.dfinity.org/t/nns-updates-aggregation-thread/23551). 

## Neuron attributes
Before explaining how to modify a neuron, it is important to understand the most important parts of a neuron.
Every user can stake ICP utility tokens into a neuron and pariticipate in the NNS DAO with this neuron.  

* **Neuron ID**: Each neuron has an identity selected by NNS governance at neuron creation.
* **Account**: Each neuron has an associated account on the ICP ledger where the locked ICP balance resides. This account is owned by the NNS governance canister and therefore a user cannot move staked tokens.

> * **Controller**: The principal that controls the neuron. The principal must identify a public key pair, which acts as a “master key,” such that the corresponding secret key should be kept very secure. A principal can control many neurons.

Reviewer 2: ```suggestion
* **Controller**: The principal that controls the neuron. A principal corresponds to the public key of a public-private keypair, and anyone with possession of the corresponding private key will have full control of the neuron. Therefore, the private key of the principal ID specified as the controller should be kept very secure. A principal can control many neurons.
```
Just a thought on a different phrasing, no change required

Author: Adpted a slight version of this. Thanks for the suggestion

> * **Dissolve delay & dissolve state**: The tokens in a neuron are locked for a specified time, called the _dissolve delay_. This can be thought of like a kitchen timer that can only be turned in one direction. It can be arbitrarily increased, but only reduced by turning on the countdown and waiting for the time to pass. A neuron is eligible to vote if it has a dissolve delay of at least 6 months.

Reviewer 2: ```suggestion
* **Dissolve delay & dissolve state**: The tokens in a neuron are locked for a specified time, called the _dissolve delay_. This can be thought of like a kitchen timer that can only be turned in one direction. It can be arbitrarily increased, but only reduced by turning on the countdown and waiting for the time to pass. A neuron must have a dissolve delay of at least 6 months to be eligible to vote.
```
reworded the last sentece since having a dissolve delay of 6 months is not technically a sufficient condition to be eligible to vote

  * A neuron can be _non-dissolving_ which means that the timer is stopped and the neuron's dissolve delay remains the same. 
  * A neuron can be _dissolving_ which means that the timer is decreasing the neuron's dissolve delay as time passes. 
  * Once the timer has counted down, a neuron is _dissolved_ and the ICP tokens can be unstaked again.
* **Age**: Every non-dissolving neuron has an age, which denotes how long it has been in the non-dissolving state. 

> A neuron's dissolve delay and age increases the neuron's voting power. 

Reviewer 2: ```suggestion
A neuron's voting power increases as its age increases. A neuron's voting power also increases if its controller increases its dissolve delay.
```
no change required, just thought this was clearer

* **Maturity**:  When neurons vote, over time the NNS increases their maturity to reward them. Maturity can be converted into ICP by spawning (see below).

## Interacting with neurons
All interactions with a neuron go through the following API on the NNS governance canister. This includes everything in a neuron's lifetime such as the neuron's creation, modifications and voting, and unstaking the ICP tokens.
```
type ManageNeuron = record {
  id : opt NeuronId;
  command : opt Command;
  neuron_id_or_subaccount : opt NeuronIdOrSubaccount;
};
```

### Creating a neuron and topping it up
Staking ICP utility tokens in a neuron involves three steps:

> 1. Compute the neuron's address on the ledger, where the staked tokens will be held. This corresponds to a subaccount of the NNS governance canister.

Reviewer 2: NCR

we seriously should make this step easier I think. It would be great to have a website or a governance API that did this computation for you given a principal ID and a nonce

2. Send ICP utility tokens to the neuron's address computed in (1).
3. Claim the neuron on the NNS governance canister, which tells the governance canister that the transfer in (2) happened, upon which the NNS governance will create a new neuron.

Let us go through these steps in more detail.

**Compute the neuron's address**

To compute the neuron's address in the first Step, proceed as follows.

> a) Learn the principal that should control the neuron and choose a nonce.

Reviewer 2: ```suggestion
a) Learn the principal that should control the neuron and choose a nonce. (The nonce is used to allow the same principal to control multiple neurons and does not need to be kept secret.)
```
BTW, is it called a nonce or a memo?

Author: I think the problem is that we are inconsistent. I would say "nonce" is clearer (a number only used once is what we want here), but the ledger field that defines this for neurons is called "memo" - the memo field is more general than this, but I think for this reason we eneded up calling it "memo" in some places

It is recommended to test the corresponding code thoroughly. 

:::

**Send ICP to the neuron's account**

After having computed the neuron's address, make a transfer to this address. For this, view the [relevant `candid` interface of the ICP ledger](https://github.com/dfinity/ic/blob/master/rs/rosetta-api/icp_ledger/ledger.did#L25) which contains the following details for a transaction.
```
type Transaction = record {
    memo : Memo;
    icrc1_memo: opt blob;
    operation : opt Operation;
    created_at_time : TimeStamp;
}; 
type Operation = variant {
    ...
    Transfer : record {
        from : AccountIdentifier;
        to : AccountIdentifier;
        amount : Tokens;
        fee : Tokens;
        spender : opt vec nat8;
    };
    ...
};
```
To make a transfer to the correct account, choose the `memo` equal your chosen nonce (from Step (1a)) and `operation` to be a transfer. You can make the transfer from any `from` account that you control. For the `to` account, use the account that you have computed in Step (1).
To later stake a neuron, the `amount` must be at least 1 ICP. Set the `fee` to the standard fee of the ICP ledger canister.

> 

Author: Do you mean that the above (where I point to the candid) is the wrong endpoint or that you find some of what follows not necessary?

> 

Author: I now pointed to the endpoint that you mentioned. Do you think this is OK now? What should create_at_time be set to? to now?

> 

Author: @bogwar I would like to merge this PR now as it has been open for a while. If you have follow up concerns, please let me know and I can add stuff in a new PR

**Claim the neuron**

As a third step, claim the neuron using the following `ManageNeuron` command.
```
type ClaimOrRefreshNeuronFromAccount = record {
  controller : opt principal;
  memo : nat64;
};
```
The `controller` is the principal controller chosen in Step (1a) that will control the neuron and the `memo` is again the chosen nonce. 

_Required permissions:_ Anyone can make this call and claim a neuron for the specified controller.

**Topping up a neuron**
Topping up an existing neuron with more tokens is called _refreshing_ a neuron and works similarly to claiming a neuron.
1. Send additional ICP utility tokens to the (existing) neuron's address.
2. Refresh the neuron on the NNS governance canister, which tells the governance canister that the transfer in (1) happened, upon which the NNS governance will update the neuron's stake. 

:::info

> Note that refreshing of a neuron will also update the neuron's age to account for the fact that the newly added tokens have no age.

Reviewer 2: ```suggestion
Note that refreshing of a neuron will also reduce the neuron's age to account for the fact that the newly added tokens have no age.
```

:::

These steps happen analogously to the claiming of neurons above.
Again, care is advised to makes sure that the right account is computed and used.
When `ClaimOrRefreshNeuronFromAccount` is used, the NNS governance canister will automatically either claim the neuron if it doesn't exist yet or top it up if it exists.

_Required permissions:_ Anyone can make this call and top up a neuron, even if they do not have any permissions on the neuron. No matter who makes the call to the governance canister, its controller is set by the arguments as mentioned above.

### Modifying a neuron's state

> Recall from the section _Neuron attributes_, that a neuron has a dissolve delay and can be non-dissolving, dissolving, or dissolved.
> To switch between the diffent dissolve states or to increase the dissolve delay, use the `ManageNeuron` command `Configure`. 

Reviewer 4: ```suggestion
Recall from the section _Neuron attributes_, that a neuron has a dissolve delay and can be non-dissolving, dissolving, or dissolved.
To switch between the different dissolve states or to increase the dissolve delay, use the `ManageNeuron` command `Configure`. 
```

To move a neuron from dissolving into non-dissolving state, and stop the timer at the dissolve state that the neuron has, use `StopDissolving`.

No matter in which dissolve state a neuron is, its dissolve delay can be increased up to a maximum of eight years.
To do so, there are two options.
You can use `IncreaseDissolveDelay` and specify by the argument `additional_dissolve_delay_seconds` by how many seconds the current remaining dissolve delay should be increased. For example, if a neuron has 6 months dissolve delay, choosing the argument 2 months (in seconds) for `IncreaseDissolveDelay`, would result in a neuron with 8 months dissolve delay. 
Alternatively, you can specify how much dissolve delay a neuron should have after the operation by using `SetDissolveTimestamp` and setting a target time stamp `dissolve_timestamp_seconds` in seconds from the Unix epoch.
To achieve the same as above, compute the timestamp of now plus 8 months and provide this as the input `dissolve_timestamp_seconds`. 
In both cases, the dissolve delay can only be increased but never decreased.

> 

Reviewer 4: This is a bit funky and counterintuitive. In particular we're using this wrongly in Rosetta (afaiu) @MarioDfinity

Author: Is this just a comment for Mario to check in Rosetta but nothing to change in the text here?

The new neuron can remain controlled by the same principal as the parent neuron, or be assigned to a new principal.

The following command spawns a neuron.
```
type Spawn = record {
  percentage_to_spawn : opt nat32;
  new_controller : opt principal;
  nonce : opt nat64;
};
```
Specify by `perpercentage_to_spawn` the portion of the maturity that should be spawn into a neuron in percentage. This should be a value between 1 and 100 (inclusive). 
If the spawned neurons should have a different controller than the parent neuron (from which you spawn the maturity), you can optionally define a different controlling principal by `new_controller`. Otherwise, the spawned neuron will have the same controller as the parent neuron. 

> Similarly to claiming a neuron (see above), you need to specify the new neuron's ledger address and choose a nonce for this in the last argument `nonce`.

Reviewer 4: What happens if the nonce is not specified? What if it specifies an already existing neuron?

Author: In that case it seems that a random nonce is generated : https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/nns/governance/src/governance.rs?L2906

Good point - I will add this!

Based on this nonce and the controlling principal, the neuron's address is uniquely defined (see section _Compute the neuron's address_).

_Required permissions:_
Only a neuron's controller can spawn maturity from a neuron.

### Disbursing / unstaking a neuron
When a neuron is dissolved, i.e., its dissolve delay is zero, the neuron's controlling principal can instruct it to _disburse_ the neuron’s stake. 

> This means that the staked ICP balance is transferred to a specified new ledger account and is liquid again.

Reviewer 4: ```suggestion
This means that the staked ICP balance is transferred to a specified ledger account and is liquid again.
```

There are two topics that are excluded from the catch-all: `Governance` and `SNS & Neuron's Fund`. 

:::info
It is important to understand that a neuron's vote is only cast if the majority of the follwed neurons agree on a decision. Find [here](https://internetcomputer.org/docs/current/developer-docs/daos/nns/nns-app-quickstart#neuron-following) more information and examples regarding the following-rules.

:::

To define a follow rule, use the following neuron command
 ```
type Follow = record {
  topic : int32; 
  followees : vec NeuronId 
};
```
The argument `topic` defines for which proposal topic this following rule should be applied.
If the topic is `null`, then this rule is applied for the catch-all and will be applied for all topics where no other explicit rule is defined except for`Governance` and `SNS & Neuron's Fund`.
The second argument `followees` defines a list of neuron IDs whose decision will be followed. There can be up to 15 followees for each topic (but more followees is not necessarily better - see the rules linked above).

_Required permissions:_
Voting and setting following can be done by a neuron's controller or a neuron's hot key.

> ### Managing permissions of a neuron

Reviewer 4: The information on hotkeys should be moved up, since otherwise the _Required permissions_ is less informative if it's not clear what it refers to. 

Also, I missed this: how can one see information about the neuron and what is it that the public can see and the owner can see?

Author: Moved this section up (right after creating of a neuron).
Will add a section on reading the neuron too.
